README.md
ReallifeRedstone
======================
0.0.2

Bukkit Plugin to breakout redstone signals to a mqtt server.
this code is still very much in alpha state. 

Mosquitto( http://mosquitto.org/ ) is a nice example of an mqtt server that's multiplatform
if you are on a pi,"sudo apt-get install mosquitto" will install a server, but be aware, you 
are now running a MQTT server, and you might not want this open to all.
  
The code for this plugin is available at: https://github.com/ajtag/ReallifeRedstone 
For more information, try: http://electrictea.co.uk/pimc

Paho is a MQTT library from the Eclipse foundation, distributed under the EPL licence,
more information is available at their website: http://www.eclipse.org/paho/



INSTALLING
==========
unzip folder, and copy contents of MCMQTT folder into craftbukkit plugins folder.
run craftbukkit server


USAGE
==========
mqttconnect [server]
	To connect to mqttserver and start sending updates 

mqttconnect 
	to save you entering the details for your server on every connection, 
	edit McMqtt/config.yml file to have your mqtt server's address there
	currently defaulting to loacalhost.
 
mqttdisconnect  
	disconnect from server


to create an IO point, place a sign, the first line defines the direction of data.
The second line is the signid (you are best using alpha or numbers only for the moment). 

	"[input]" creates a input point (don't type in the " marks)
	on placing this an input sign it turns into a lever, that players cannot modify, only destroy.
	the lever can be modified by sending a message to the topic that the sign listens on.
	
	to update it's state, send the string "1" or "0" to  minecraft/input/*signid* 
	Edit *signid* as appropriate.
	
	
	"[output]" creates a sign that sends a message whenever there is a change of redstone signal on the sign.
	it will send "0" for no signal or "15" if there is a signal.
	
	the signal will be sent to minecraft/output/*signid*
	  

Pi_scripts
==========
	Contains code designed to run on a raspberry Pi to push mqtt messages out
	onto the PiFace board.
	this code assumees you have the python-mosquitto and piface(pfio) librarys available.
	copy this onto your Pi and run with your python interpreter.

Docs
==========
	Contains EPL.txt, the licence that covers the Paho library.
	and LGPL.txt, the licence text to cover this code.
	
Paho is a MQTT library from the eclipse foundation, distributed under the EPL licence,
more information is available at their website: http://www.eclipse.org/paho/


Angus Taggart, 2013
	

Changelog.txt
v0.0.2
	Fixed Bug missing init of array.

v0.0.1
	Initial Release

pi_script/pfmqtt.py
#! /usr/bin/python
'''
apt-get install python-mosquitto
this assumes you have your piface installed and the python library installed too
'''

import mosquitto
import piface.pfio as pfio
import re


class pfmqtt:
    def __init__(self, ident, topicheader, uri = 'localhost', port = 1883):
        self.uri = uri
        self.port = port
        self.ident = ident
        
        self.topicheader = topicheader+ '/' + self.ident
        self.outputRegex = re.compile(self.topicheader + "/output/pin/(\d*)", re.IGNORECASE)
        
        
        self.mos = mosquitto.Mosquitto(self.ident)
        self.mos.on_message = self.on_message # register for callback
        self.connect()
        
        
        pfio.init()
        self.inputpins = pfio.read_input()
        
    def connect(self):
        conn = self.mos.connect(self.uri, port=self.port)
        if conn != 0:
            print 'Connection Failed with error: ' + str(conn)
            raise "CONNECTION ERROR"    
        self.mos.subscribe(self.topicheader + "/output/pin/+", 0) # get all messages for me

    def run(self):
        while self.mos.loop(0) == 0:
            inputpins = pfio.read_input() # get input pin state
            if self.inputpins != inputpins: # check to see if they are different
                change = self.inputpins ^ inputpins #get the differences
                for i in range(8): 
                    pinid = 1<<i
                    if change & pinid == pinid:
                        self.mos.publish( self.topicheader + '/input/pin' + str(i), str((inputpins&pinid)>>i), 0 )
                self.inputpins = inputpins
                
    def on_message(self,  obj, msg):
        
        rm = self.outputRegex.match( msg.topic )
        
        if rm != None:
            pin = int(rm.group(1))
            #print msg.topic + " - " + msg.payload
            
            if msg.payload == '0':
                pfio.digital_write(pin, 0)
                #print "setting pin %d to %d"%(int(rm.group(1)), 1)
            else:
                pfio.digital_write(pin, 1)
                #print "setting pin %d to %d"%(int(rm.group(1)), 0)
                
    
    def __del__(self):
        self.mos.disconnect()
        pfio.deinit()
if __name__ == '__main__':
    conn = pfmqtt('pi', 'minecraft', 'localhost', 1883)
    conn.run()


    

src/uk/co/electrictea/mc_mqtt/McMqtt.java
/*
 	Copyright 2013 Angus Taggart
	
	This file is part of mcmqtt.

    mcmqtt is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    mcmqtt is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser Public License for more details.

    You should have received a copy of the GNU Lesser Public License
    along with mcmqtt.  If not, see <http://www.gnu.org/licenses/>.
*/
package uk.co.electrictea.mc_mqtt;



import java.util.ArrayList;

import org.bukkit.block.Block;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
//import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.plugin.java.JavaPlugin;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttSecurityException;


interface BlockStore{
	void addBlock(Block ioBlock, String[] signText);
	void removeBlock(Block ioBlock);
	boolean isIoBlock(Block ioBlock);
	ArrayList<InputLever> getInputBlockLocations(String topic);
	int length()
	;
}

public final class McMqtt extends JavaPlugin {
	enum BlockStorageTypes{Memory};
	BlockStore ioBlockStore;
	
	private Courier cl = null;
	private McIO mc = null;
	
		
    @Override
    public void onEnable(){
		saveDefaultConfig();
		switch(BlockStorageTypes.valueOf(getConfig().getString("BlockStore.Type"))){
    	default:
    		getLogger().info("StoreType not recognised, using: Memory");
    	case Memory:
    		ioBlockStore = new MemoryStore();
    	}		
    	mc = new McIO(this, ioBlockStore);
    	
    	getServer().getPluginManager().registerEvents(this.mc, this);
    	getLogger().info("mcmqtt Enabled!");
    }
    
    @Override
    public void onDisable() {
    	getLogger().info("mcmqtt Disabled!");
    	try {
			cl.disconnect();
		} catch (MqttException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
    }
    
    @Override    
    public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
    	if (cmd.getName().equalsIgnoreCase("mqttconnect")){ 
    		
    		
    		try {
	    		if (args.length == 0){
	    			cl = new Courier(getConfig().getString("Mqtt.Server"), getConfig().getString("Mqtt.ClientID"), getConfig().getString("Mqtt.Topic"), mc);	
	    		}else if (args.length == 1){
	    			cl = new Courier(args[0].toString()					 , getConfig().getString("Mqtt.ClientID"), getConfig().getString("Mqtt.Topic"), mc);	
	    		}
	    		cl.connect();
	  
    		} catch (MqttSecurityException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			} catch (MqttException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
    		return true;
    	} if (cmd.getName().equalsIgnoreCase("mqttdisconnect")){
    		try {
				this.cl.disconnect();
			} catch (MqttException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
    		return true;
    	}
    		
    	return false;
    }
    
    public void publish(byte[] payload, String topicName){
    	if (cl != null){
    		if (cl.isConnected()){
    			cl.pubm(payload, topicName, 0);
    		}
    	}
    }
}


src/uk/co/electrictea/mc_mqtt/Courier.java
/*
 	Copyright 2013 Angus Taggart
	
	This file is part of mcmqtt.

    mcmqtt is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    mcmqtt is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser Public License for more details.

    You should have received a copy of the GNU Lesser Public License
    along with mcmqtt.  If not, see <http://www.gnu.org/licenses/>.
*/
package uk.co.electrictea.mc_mqtt;

import java.util.logging.Logger;

import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttCallback;
//import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.eclipse.paho.client.mqttv3.MqttPersistenceException;
import org.eclipse.paho.client.mqttv3.MqttSecurityException;
import org.eclipse.paho.client.mqttv3.MqttTopic;

import uk.co.electrictea.mc_mqtt.McIO;

public class Courier extends MqttClient implements MqttCallback{
	private String id;
	private String topic;
	private McIO mcIO; 
	
	public Courier(String URI, String id, String topic, McIO mcIO) throws MqttException{
		super("tcp://"+ URI + ":1883", id);
		Logger.getLogger("info").info("Using MQTT server: '" + "tcp://"+ URI + ":1883" + "' with ClientID '" + this.id + "'.");
		
		this.mcIO  = mcIO;
		this.topic = topic;
		this.id = id;
	}
	
	@Override
	public void connect() throws MqttSecurityException, MqttException{
		super.connect();
		//subscribe to all input messages related to our config
		setCallback(this);
		subm(topic+"/input/#"); 
	}

	 public void connectionLost(java.lang.Throwable arg0){
		  Logger.getLogger("INFO").info("Connection to MQTT Lost");
		  try {
			connect();
		  } catch (MqttSecurityException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		  } catch (MqttException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	 }
	 
	  

	  public void messageArrived(MqttTopic arg0, MqttMessage arg1) throws java.lang.Exception {
		  if (arg0.toString().startsWith(topic+ "/input/")){
			  //Logger.getLogger("INFO").info(topic +" - " + new String(arg1.getPayload()));
			  mcIO.updateLevers(arg0.getName().replace(topic+ "/input/", ""), new String(arg1.getPayload()));
		  }
		  
	  }
	  
	  public void deliveryComplete(MqttDeliveryToken arg0){
		  
	  }
	
	
	
	public void pubm(byte[] payload, String topicName, int qos ){
		MqttTopic topic = null;
		try{
			topic = this.getTopic(this.topic+ "/" + topicName);
		}catch(Exception  e){
			e.printStackTrace();
		}
		MqttMessage message = new MqttMessage(payload);
		message.setQos(qos);
			
		try {
			Logger.getLogger("INFO").info(new String(payload));
			//MqttDeliveryToken token = 
			topic.publish(message);
		} catch (MqttPersistenceException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (MqttException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	
	}
	
	public void subm(String topicname){
    	try {
			this.subscribe(this.topic+ "/input/"+topicname);
		} catch (MqttSecurityException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (MqttException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
	
	
}


src/uk/co/electrictea/mc_mqtt/InputLever.java
/*
 	Copyright 2013 Angus Taggart
	
	This file is part of mcmqtt.

    mcmqtt is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    mcmqtt is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser Public License for more details.

    You should have received a copy of the GNU Lesser Public License
    along with mcmqtt.  If not, see <http://www.gnu.org/licenses/>.
*/
package uk.co.electrictea.mc_mqtt;

import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.block.Block;

public class InputLever {
	private String[] signText;
	private Location signLocation;
	private World world;
	
	public World getWorld(){
		return world;
	}
	
	public InputLever(String[] signText, Location signLocation, World signWorld){
		this.signText = signText;
		this.signLocation = signLocation;
		this.world = signWorld;
	}
	
	public Location getLocation(){
		return this.signLocation;
	}
	
	public String getSignType(){
		return this.signText[0].replace("[", "").replace("]", "");
	}
	public String[] getSignText(){
		return this.signText;
	}
	public String getSignName(){
		return this.signText[1];
	}
	public Block getBlock(){
		return world.getBlockAt(signLocation);
	}
}

src/uk/co/electrictea/mc_mqtt/MemoryStore.java
/*
 	Copyright 2013 Angus Taggart
	
	This file is part of mcmqtt.

    mcmqtt is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    mcmqtt is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser Public License for more details.

    You should have received a copy of the GNU Lesser Public License
    along with mcmqtt.  If not, see <http://www.gnu.org/licenses/>.
*/
package uk.co.electrictea.mc_mqtt;

import java.util.ArrayList;


import org.bukkit.Material;
import org.bukkit.block.Block;

public class MemoryStore implements BlockStore{
	//this store is temporary, so Input levers will only last as long as the server
	ArrayList<InputLever> inputBlocks;
	
	public MemoryStore () {
		inputBlocks = new ArrayList<InputLever>(10);
	}
	
	public int length(){
		return inputBlocks.size();
	}
	
	public void addBlock(Block ioBlock, String[] signText) {
		this.inputBlocks.add(
				new InputLever(
						signText, 
						ioBlock.getLocation(), 
						ioBlock.getWorld()
				)
			);
	}
	
	public boolean isIoBlock(Block ioBlock){
		if(ioBlock.getType() == Material.LEVER){
			for(int i=0; i < this.inputBlocks.size(); i++){
				if (this.inputBlocks.get(i).getLocation().hashCode() == ioBlock.getLocation().hashCode()){
					return true;
				}
			}
		}
		return false;
	}
	
	public ArrayList<InputLever> getInputBlockLocations(String topic){
		ArrayList<InputLever> bl = new ArrayList<InputLever>();
		for(int i=0; i < this.inputBlocks.size(); i++){
			if (inputBlocks.get(i).getSignName().equalsIgnoreCase( topic ) ){
				bl.add(inputBlocks.get(i));
			}	
		}
		return bl;
	}
	
	public void removeBlock(Block ioBlock){
		if(ioBlock.getType() == Material.LEVER){
			for(int i=0; i < this.inputBlocks.size(); i++){
				if (this.inputBlocks.get(i).getLocation().hashCode() == ioBlock.getLocation().hashCode()){
					this.inputBlocks.remove(i);
				}
			}
		}
	}
}

src/uk/co/electrictea/mc_mqtt/switchListener.java
/*
 	Copyright 2013 Angus Taggart
	
	This file is part of mcmqtt.

    mcmqtt is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    mcmqtt is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser Public License for more details.

    You should have received a copy of the GNU Lesser Public License
    along with mcmqtt.  If not, see <http://www.gnu.org/licenses/>.
*/
package uk.co.electrictea.mc_mqtt;

import org.bukkit.Material;
import org.bukkit.event.Listener;
import org.bukkit.event.EventHandler;

import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockRedstoneEvent;
import org.bukkit.event.block.SignChangeEvent;
import org.bukkit.block.Sign;

import java.util.logging.Logger;


public class switchListener implements Listener {
	 @EventHandler
	 public void BlockRedstoneEvent(BlockRedstoneEvent bre){
		//Location breloc;
		//breloc = bre.getBlock().getLocation();
		Logger.getLogger("REDSTONE").info( "" + bre.getBlock().getTypeId() );
		
		if (  /*(bre.getBlock().isBlockPowered() == true) &&*/ ((bre.getBlock().getTypeId() == 63) || bre.getBlock().getTypeId() == 68) ){
			Logger.getLogger("REDSTONE").info( ((Sign) bre.getBlock().getState()).getLine(0));
		}
	 }


	 
	 @EventHandler
	 public void SignChange(SignChangeEvent bce){
		 if (((bce.getBlock().getTypeId() == 63) || bce.getBlock().getTypeId() == 68)){
			 if ( bce.getLine(0).equalsIgnoreCase("[output]")){
				 Logger.getLogger("INFO").info("Output Sign");
				 byte lever = 0x0;
				 
				 if (bce.getLine(1).equalsIgnoreCase("1")){
					 Logger.getLogger("INFO").info("Sign HIGH");
					 lever = (byte) (lever | 0x8);
				 }else{
					 Logger.getLogger("INFO").info("Sign LOW");
					 lever = (byte) (lever & ~ 0x8);
				 }
					 
					 if (bce.getBlock().getTypeId() == 63){
						 lever = (byte) (lever | 0x5 );
					 }else{
						 byte wallsign = bce.getBlock().getData();
						 if ((wallsign & 0x7) == 0x2) //north
							 lever = (byte) (lever | 0x4);
						 if ((wallsign & 0x7) == 0x4) //west
							 lever = (byte) (lever | 0x2);
						 if ((wallsign & 0x7) == 0x3) //east
							 lever = (byte) (lever | 0x3);
						 if ((wallsign & 0x7) == 0x5) //south
							 lever = (byte) (lever | 0x1);
					 }	 
					 bce.getBlock().setType(Material.LEVER);
					 bce.getBlock().setData(lever);	 
					 if ((lever & 0x8) == 0x8){
						 ;
					 }
				}
			 }
		 }	 
	 
	 
	 @EventHandler
	 public void blockDestroy(BlockBreakEvent bbe){
		 if (bbe.getBlock().getType().getId() == 69){
			 Logger.getLogger("INFO").info("Lever Destroyed!");
		 }
	 }
}



src/uk/co/electrictea/mc_mqtt/McIO.java
/*
 	Copyright 2013 Angus Taggart
	
	This file is part of mcmqtt.

    mcmqtt is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    mcmqtt is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser Public License for more details.

    You should have received a copy of the GNU Lesser Public License
    along with mcmqtt.  If not, see <http://www.gnu.org/licenses/>.
*/
package uk.co.electrictea.mc_mqtt;

import java.util.ArrayList;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.Sign;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockRedstoneEvent;
import org.bukkit.event.block.SignChangeEvent;
import org.bukkit.event.player.PlayerInteractEvent;

public class McIO implements Listener {
	private McMqtt mcmqtt;
	private BlockStore ioBlockStore;
	
	public McIO(McMqtt mcmqtt, BlockStore ioBlockStore){
		this.mcmqtt = mcmqtt;
		this.ioBlockStore = ioBlockStore;
	}
	
	
	public void updateLevers(String topic, String message){
	  //TODO: rewrite this to return a list of matching blocks
	  //int blockid = this.getInputBlockID( topic );
	  //Logger.getLogger("INFO").info("BID: " +blockid);
		Block ib;
		ArrayList<InputLever> bl  = ioBlockStore.getInputBlockLocations(topic);
		int i = bl.size();
		byte lever;
		
		while (i > 0){
			ib = bl.get(i).getBlock();		
			//bl.get(i).getWorld().getBlockAt(bl.get(i).getLocation());
			lever = ib.getData();
			if(ib != null){
				  if( message.equalsIgnoreCase("1")){
					// Logger.getLogger("INFO").info("Set HIGH");
					 lever = (byte) (lever | 0x8);
				  }else{
					// Logger.getLogger("INFO").info("Set LOW");
					 lever = (byte) (lever & ~ 0x8);
				  }
				  ib.setData(lever);
				  updateLeverBlock(ib);		
			  }
			i--;
		}
	}
	
	
	
	@EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
	public void onInteract(PlayerInteractEvent event){
		if (event.getAction() == Action.RIGHT_CLICK_BLOCK || event.getAction() == Action.RIGHT_CLICK_AIR)
		{
			if( ioBlockStore.isIoBlock( event.getClickedBlock() ) ){
				event.setCancelled(true);
				event.getPlayer().chat("you can't do that Hal"); 
			}
		}
    }

	@EventHandler //Priority
	 public void BlockRedstoneEvent(BlockRedstoneEvent bre){
		/*
		 * code for 2 way levers
		 */
		/*
		if (bre.getBlock().getType() == Material.LEVER){
			String lstat;
			if ((bre.getBlock().getData() & 0x8) != 0x8){ lstat = "on"; }else{ lstat = "off"; }
			if(this.isInputBlock(bre.getBlock())){
					Logger.getLogger("INFO").info("Got Input Block");
			}
			//this.pubm( ("lever:"+lstat).getBytes() , "Lever/"+bre.getBlock().getLocation().toString(), 0);
		}
		*/
		
		
		if ( ( (bre.getBlock().getType() == Material.SIGN) || bre.getBlock().getType() == Material.SIGN_POST) ){
			String[] signText = ((Sign) bre.getBlock().getState()).getLines();
			if (signText[0].equalsIgnoreCase("[output]")){
				String signid = signText[1];
				byte[] message = {(byte)bre.getNewCurrent()};//new byte[1];
				
				
				this.mcmqtt.publish((String.valueOf((int)message[0])).getBytes() , "output/"+ signid);		
			}
		}
		
	 }
	
	 
	 @EventHandler
	 public void BreakBlock(BlockBreakEvent bbe){
		 ioBlockStore.removeBlock(bbe.getBlock());
	 }
	 
	 @EventHandler
	 public void SignChange(SignChangeEvent bce){
		 if (((bce.getBlock().getTypeId() == 63) || bce.getBlock().getTypeId() == 68)){
			 if ( bce.getLine(0).equalsIgnoreCase("[input]")){
				 //Logger.getLogger("INFO").info(bce.getLines()[0]);
				 //Logger.getLogger("INFO").info(bce.getBlock);
				 ioBlockStore.addBlock(bce.getBlock(), bce.getLines());
				 byte lever = 0x0;
					 if (bce.getBlock().getTypeId() == 63){
						 lever = (byte) (lever | 0x5 );
					 }else{
						 byte wallsign = bce.getBlock().getData();
						 if ((wallsign & 0x7) == 0x2) //north
							 lever = (byte) (lever | 0x4);
						 if ((wallsign & 0x7) == 0x4) //west
							 lever = (byte) (lever | 0x2);
						 if ((wallsign & 0x7) == 0x3) //east
							 lever = (byte) (lever | 0x3);
						 if ((wallsign & 0x7) == 0x5) //south
							 lever = (byte) (lever | 0x1);
					 }	 
					 bce.getBlock().setType(Material.LEVER);
					 bce.getBlock().setData(lever);	 
				}
			 }
		 }	
	
		
		/*private Block getInputBlock(int inputID){
			return this.world.getBlockAt(this.inputBlocks.get(inputID).getLocation());
		}
		*/

	/*
		private int getInputBlockID(String inputID){
			Logger.getLogger("INFO").info("Searching for :" + inputID);
			for(int i=0; i < this.inputBlocks.size(); i++){
				if (this.inputBlocks.get(i).getSignName().equalsIgnoreCase( inputID ) ){
					return i;
				}else
				 Logger.getLogger("INFO").info(this.inputBlocks.get(i).getSignName());
				
			}
			return -1;
		}
*/
		private void updateLeverBlock(Block lb){
			
			 if (lb.getType() != Material.LEVER){
				 return;
			 }
				 
				 byte ld = lb.getData();
				 Location lloc = lb.getLocation();
				 Block b = lb;
				 
				 if ((ld & 0x7) == 0x7 || (ld & 0x7) == 0x0) //ceiling
					 b = lloc.add(0, 1, 0).getBlock();
				 else if ((ld & 0x7) == 0x5 || (ld & 0x7) == 0x6) //floor
					 b = lloc.add(0, -1, 0).getBlock();
				 else if ((ld & 0x7) == 0x4) //north
					 b = lloc.add(0, 0, -1).getBlock();
				 else if ((ld & 0x7) == 0x2) //west
					 b = lloc.add(-1, 0, 0).getBlock();
				 else if ((ld & 0x7) == 0x3) //east
					 b = lloc.add(1, 0, 0).getBlock();
				 else if ((ld & 0x7) == 0x1) //south
					 b = lloc.add(0, 0, 1).getBlock();
				 
				 byte oldData = b.getData();
				 byte notData;
				 
				 if (oldData>1) notData = (byte)(oldData-1);
				 else if (oldData<15) notData = (byte)(oldData+1);
				 else notData = 0;
				 
				 b.setData(notData, true);
				 b.setData(oldData, true);

		}
		
}
